openhcl/fixed_pool_alloc/src/lib.rs

```diff
@@ -0,0 +1,573 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+// Copyright (C) Microsoft Corporation. All rights reserved.
+
+//! This module implements a fixed memory allocator for allocating pages at specific location.
+
+#![cfg(unix)]
+#![warn(missing_docs)]
+// SAFETY: Send, Sync, and *nix calls mmap() munmap() require unsafe keyword.
+#![allow(unsafe_code)]
+
+mod mapped_dma;
+
+pub use mapped_dma::FixedDmaBuffer;
+
+use anyhow::Context;
+use hvdef::HV_PAGE_SIZE;
+use inspect::Inspect;
+use memory_range::MemoryRange;
+use parking_lot::Mutex;
+use std::num::NonZeroU64;
+use std::sync::Arc;
+use thiserror::Error;
+use user_driver::memory::MemoryBlock;
+use user_driver::vfio::VfioDmaBuffer;
+use user_driver::HostDmaAllocator;
+
+/// Error returned when unable to allocate memory.
+#[derive(Debug, Error)]
+#[error("unable to allocate fixed pool size {size} with tag {tag}")]
+pub struct FixedPoolOutOfMemory {
+    size: u64,
+    tag: String,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, Inspect)]
+#[inspect(external_tag)]
+enum State {
+    Free {
+        #[inspect(hex)]
+        base_pfn: u64,
+        #[inspect(hex)]
+        size_pages: u64,
+    },
+    Allocated {
```

Comment:
I am a bit worried that we will fail to restore some allocations and will never notice until we hit memory corruption. I also don't see a mechanism where we prevent the allocator from handing out saved allocations, e.g. if one device is allocating before another device is done restoring.

I would suggest we improve the logic like this:
1. Add a parameter to `FixedPool::allocator` for whether this specific allocator is for a device that will be saved/restored.
2. When allocating via a save/restore-capable allocator, remember that fact in `State`.
3. As part of servicing state, save the list of ranges that were allocated as part of saved devices.
4. When recreating the allocator on restore, create entries for each saved range.
5. Make sure `restore` can only restore a range that was previously saved (at which point you convert it to an allocated range). Make sure `alloc` never allocates on top of a saved range.
6. After all device initialization has completed, we should be able to assert that there are no remaining saved ranges. They should all be converted to allocated ranges by this point. We can fail VTL0 start if there are still ranges (or we can leak the memory forever, policy decision).

Reply:
What I am getting from this - you propose to move save/restore memory logic from NVMe queues to the allocator itself, and restore queues with either already-present MemBlock, or validate that the queues restored to the same range as allocator?

Reply:
Added save/restore to the allocator (will be superseded by upcoming @chris-oo change)